[MO] fixed dealing with None in then/else branches of Select

pavel-esir opened the pull request:
### Details:
 **root cause analysis:** In some cases `Select` condition is True[False] for all elements while one of the branches is None (which is not selected). 
During shape inference of `Select` `np.where` is called and in the above case output_value `dtype` is set to `object`. Which is not compatible with numerical int and float types. 
![image](https://user-images.githubusercontent.com/5703039/132849128-fa093074-e535-444d-8dcb-b8470de73fec.png)
And subsequently numpy operations were failed without proper error message and explanation.

**solution**: process cases when there is `None` in one the branches manually without using `np.where` so that `dtype` will not be corrupted
![image](https://user-images.githubusercontent.com/5703039/132848658-2132ba66-5b3a-40be-97e7-f15094e4b170.png)

### Tickets:
 - xxx-62426, xxx-64129
 
Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: done
* [x]  Framework operation tests: no new operation were added
* [x]  Transformation tests: no new transformation were added
* [x]  Model Optimizer IR Reader check: done

Documentation:
* [x]  Supported frameworks operations list n/a
* [x]  Guide on how to convert the **public** model n/a
* [x]  User guide update: a/a

pavel-esir commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -39,27 +39,46 @@ def infer(node: Node):
+            raise Error("PDPD broadcasting rule is not implemented yet")
+        else:  # broadcasting is not allowed
+            assert np.array_equal(a_shape, b_shape) and condition_value is not None and np.array_equal(condition_value.shape, a_shape), \
+                'In node \'{}\' For Select operation when broadcasting is off all inputs must be of the same shape. ' \
```
For -> for

pavel-esir replied:
done

pavel-esir commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -39,27 +39,46 @@ def infer(node: Node):
-                resulting_tensors[1] = bi_directional_broadcasting(resulting_tensors[1], a_shape)
-            condition_value = bi_directional_broadcasting(condition_value, output_shape)
+            if resulting_tensors[0] is None or resulting_tensors[1] is None:
+                assert np.all(condition_value == condition_value[0])
```
`condition_value[0] => condition_value.item(0)` for safety

pavel-esir replied:
done

pavel-esir commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -39,27 +39,46 @@ def infer(node: Node):
+                # in some graphs Select condition is always True[False] and
+                # one of the branches is None (which is not selected)
+                # if we use np.where for such cases then dtype of output_value will be object (non numeric type)
+                # and subsequent numpy operation on a such tensors will fail
```
on a such tensors =>  on such tensors

pavel-esir replied:
done

pavel-esir commented:
@yekruglov @evolosen please take a look

evolosen commented on model-optimizer/unit_tests/extensions/ops/select_test.py:
```diff
@@ -79,6 +79,56 @@ def test_select_infer_condition_false(self):
+    # clear explanation, need to catch such cases as soon as possible
+    def test_select_infer_None_then_branch(self):
+        graph = build_graph_with_attrs(nodes_with_attrs=self.nodes, edges_with_attrs=self.edges,
+                                       update_nodes_attributes=[('condition_data', {'value': np.array([False])}),
+                                                                ('select_output', {'shape': np.array([2, 2]),
+                                                                                   'value': np.zeros((2, 2))}),
+                                                                ('then_data', {'value': None, 'shape': np.array([2, 2])})
```
please set type of array

rkazants replied:
please use `int64_array`

pavel-esir replied:
have completely rewritten these tests, and types now are defined explicitly

evolosen commented on model-optimizer/unit_tests/extensions/ops/select_test.py:
```diff
@@ -79,6 +79,56 @@ def test_select_infer_condition_false(self):
+        # We should propagate shapes and values
+        graph_ref = build_graph_with_attrs(nodes_with_attrs=self.nodes,
+                                           edges_with_attrs=self.edges,
+                                           update_nodes_attributes=[('select_output', {'shape': np.array([2, 2]),
+                                                                                       'value': np.zeros((2, 2))}),
+                                                                    ('then_data',
+                                                                     {'value': None, 'shape': np.array([2, 2])})])
```
the same comment

pavel-esir replied:
have completely rewritten these tests, and types now are defined explicitly

evolosen commented on model-optimizer/unit_tests/extensions/ops/select_test.py:
```diff
@@ -79,6 +79,56 @@ def test_select_infer_condition_false(self):
+    # clear explanation, need to catch such cases as soon as possible
+    def test_select_infer_None_else_branch(self):
+        graph = build_graph_with_attrs(nodes_with_attrs=self.nodes, edges_with_attrs=self.edges,
+                                       update_nodes_attributes=[('condition_data', {'value': np.array([True])}),
+                                                                ('select_output', {'shape': np.array([2, 2]),
+                                                                                   'value': np.ones((2, 2))}),
+                                                                (
+                                                                'else_data', {'value': None, 'shape': np.array([2, 2])})
+                                                                ])
```
the same

pavel-esir replied:
have completely rewritten these tests, and types now are defined explicitly

evolosen commented on model-optimizer/unit_tests/extensions/ops/select_test.py:
```diff
@@ -79,6 +79,56 @@ def test_select_infer_condition_false(self):
+        # We should propagate shapes and values
+        graph_ref = build_graph_with_attrs(nodes_with_attrs=self.nodes,
+                                           edges_with_attrs=self.edges,
+                                           update_nodes_attributes=[('select_output', {'shape': np.array([2, 2]),
+                                                                                       'value': np.ones((2, 2))}),
+                                                                    ('else_data',
+                                                                     {'value': None, 'shape': np.array([2, 2])})])
```
the same

pavel-esir replied:
have completely rewritten these tests, and types now are defined explicitly

evolosen commented on model-optimizer/unit_tests/extensions/ops/select_test.py:
```diff
@@ -164,15 +214,15 @@ def test_select_infer_condition_with_value(self, condition_shape, else_data_shap
                                                                                'value': else_value}),
-                                                                ('than_data', {'shape': np.array(than_data_shape),
-                                                                               'value': than_value}),
+                                                                ('then_data', {'shape': np.array(then_data_shape),
+                                                                               'value': then_value}),
```
the same

pavel-esir replied:
have completely rewritten these tests, and types now are defined explicitly

rkazants commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -3,7 +3,7 @@
 import numpy as np
 
-from mo.front.common.partial_infer.utils import is_fully_defined
+from mo.graph.graph import Error
 from mo.graph.graph import Node, Graph
```
nit: please combine these lines

pavel-esir replied:
done

rkazants commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -39,27 +39,46 @@ def infer(node: Node):
-            if resulting_tensors[1] is not None:
-                resulting_tensors[1] = bi_directional_broadcasting(resulting_tensors[1], a_shape)
-            condition_value = bi_directional_broadcasting(condition_value, output_shape)
+            if resulting_tensors[0] is None or resulting_tensors[1] is None:
```
here you need to add a check that `condition_value` contains just one element if one of branches is None

pavel-esir replied:
done

rkazants commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -39,27 +39,46 @@ def infer(node: Node):
+                # one of the branches is None (which is not selected)
+                # if we use np.where for such cases then dtype of output_value will be object (non numeric type)
+                # and subsequent numpy operation on such tensors will fail
+                output_value = resulting_tensors[not np.bool(condition_value.item(0))]
```
what if `condition_value` is a masked array and it equals `np.ma.array([True], mask=[1])` then `np.bool(condition_value.item(0))` is True. But output_value must be with undetermined values.

pavel-esir replied:
Thanks for the valuable comment! Corrected infer function so that it correctly handles masked conditions as well. Added such cases into unit-tests

rkazants reviewed (changes requested):
please address comments

pavel-esir commented:
@rkazants applied comments. Corrected for masked condition. Also i have completely rewritten unit-tests, now dtypes are defined explicitly, removed redundant, non-correct and duplicate tests.

Left a couple of comments about my changes in discussion. Please feel free to mark them resolved after you have red them.

rkazants commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -39,28 +38,48 @@ def infer(node: Node):
+            # note that additionally to output_shape resulting_tensors must be broadcasted as well
+            raise Error("PDPD broadcasting rule is not implemented yet")
+        else:  # broadcasting is not allowed
+            assert np.array_equal(a_shape, b_shape) and condition_value is not None and np.array_equal(condition_value.shape, a_shape), \
```
since `a_shape` and `b_shape` are masked array you need to use `compatible_shapes(shape1, shape2)` from utils.py

rkazants replied:
please add tests for these cases

pavel-esir replied:
used `compatible_shapes`, added test case as well

rkazants commented on model-optimizer/extensions/ops/select.py:
```diff
@@ -39,28 +38,48 @@ def infer(node: Node):
+                'In node \'{}\' for Select operation when broadcasting is off all inputs must be of the same shape. ' \
+                'But instead got: cond_shape={}, then_shape={}, else_shape={}'.format(
+                    node_name, condition_value.shape, a_shape, b_shape)
+            output_shape = a_shape
```
In case some shape has dynamic dimension and the other one has fully defined then you need to use the fully defined dimension(, assuming that dynamic dimension must be equal).
Please test cases for it

pavel-esir commented:
@rkazants applied comments. Also additionally added one more assert when `Select` was formed from TF `Where` op. In the latest case `condition` shape must follow slightly different rule (slightly different from numpy broadcasting).
1. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_ops.py#L4596-L4598
2. https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/tensorflow_tests/test_tf_Select.py#L86-L87

Tests are failing not because of my changes. Looks like a infrastructure issue. If everything is fine could you please turn-on automerge?
